비동기처리의 패턴 중 하나

콜백 패텐이 가진 단점을 보완 - 비동기 처리시점을 명확하게 표현

### **콜백 패턴의 단점**

비동 기 함수 내부의 비동기로 동작하는 코드는 비동기 함수가 종료된 이후 완료된다

-> 따라서 비동기 함수 내부의 비동기로 동작하는 코드에서 **처리결과를 외부로 반환하거나 상위 스코프의 변수에 할당하면 기대한 대로 동작하지 않는다.!**

쉽게 설명하자면 아래와 같다.

```
const get =(url)=>{
	const xhr = new XMLHttpRequst();
    xhr.open("GET",url);
    xhr.send();

    xhr.onload=()=>{
    	if(xhr.staus === 200) {
        	return JOSN.parse(xhr.response);
        }
        else{
        	console.error("~~");
        }
    }
}
const res = get("https://~~");
console.log(res) //undefined
```

첫 번째로 **get함수에는 명시적인 반환문이 없다!**

xhr.onload 이벤트 핸들러 프로퍼티에 바인딩한 이벤트 핸들러의 반환문은(**return JSON.parse 부분**)

get 함수의 반환문이 아니다. get 함수는 반환문이 생략되었음으로 암묵적으로 undefined를 반환한다.

#### **왜 이렇게 동작하는 것일까??**

1\. get이 호출되면 함수 코드를 평가하는 과정에서 **get 함수의 실행컨텍스트 생성**

2\. 실행컨텍스트 **스택에 푸시**된다.

3\. 이후 함수 코드 실행 과정에서 xhr.onload 이벤트 핸드러프로퍼티에 이벤트핸들러가 바인딩 된다.

4. **get 함수** 종료 후 **실행컨텍스트가 스택에서 팝**된다.

5\. 곧바로 **외부의 console.log가 호출**된다. console.log의 실핸 컨텍스트가 생성되어 **실행 컨텍스트 스택에 푸시**된다.

6\. 이때 서버로부터 응답이 도착하면 **xhr 객체에서 load 이벤트 발생**

7\. xhr 이벤트 핸들러는 **태스크 큐에 저장되어 대기**

8. **콜 스택이 빈다면 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다.**

9\. 이벤트 핸들러도 함수이므로 평가-> 실행 컨텍스트 생성 -> 콜 스택  푸시 -> 실행 과정을 거친다.

**따라서 xhr.onload 이벤트 핸들러가 실행되는 시점에는 콜 스택이 빈상태여야 하므로 console.log는 이미 종료된 이후다.**

**쉽게 말해서 이벤트 루프에 의해 이러한 현상이 생기는 것이다.**

이러한 이유로

비동기 함수의 처리 결과에 대한 후속 처리는 비동기 함수 내부에서 수행해야한다.

이때 **처리 결과에 대한 후속처리를 수행하는 콜백함수를 전달하는 것**이 일반적이다.

이때 후속처리를 수행하는 비동기 함수가 비동기 처리결과를 가지고 또 다시 비동기 함수를 호출해야한다면

**콜백함수가 중첩되어 복잡도가 높아지는 현상**이 발생한다.

이를 **콜백 헬**이라고한다.

### **콜백 패턴의 문제점**

1\. 에러처리가 곤란하다.

```
try {
 setTimeout(() => { throw new Error(`Error`); },1000};
} catch(e) {
 	//에러를 캐치하지 못한다.
    console.error("캐치한 에러",e);
}
```

#### **왜 에러를 캐치하지 못할까?**

1\. 비동기 함수인 setTimeout이 호출되면 실행컨텍스트가 생성되고, 콜 스택에 푸시되어 실행 된다.

2\. 비동기 함수이므로 콜백함수가 실행되는 것을 기다리지 않고 즉시 종료되어 콜 스택에서 제거된다.

3\. 이후 타이머가 만료되면 setTimout 함수의 **콜백함수**는 **태스크 큐로 푸시**되고

4. **콜 스택이 비어졌을 때** 이벤트 루프에 의해 **콜 스택으로 푸시되어 실행**된다.

5\. 콜백 함수가 실행될 때 이미 setTimeout 함수는 이미 콜 스택에서 제거된 상태

6\. 이는 **콜백함수를 호출 한 것이 setTimeout함수가 아니라는 것을 의미**

7\. 콜백함수가 setTimeout의 함수라면 콜 스택의 **현재 실행 중인 실행 컨텍스트가 콜백 함수의 실행 컨텍스트일 때**  현재 실행 중인 **실행 컨텍스트의 하위 실행 컨텍스트가 setTimeout 함수여야한다.**

**에러는 호출자 방향으로 전파**된다.

하지만 위 처럼 콜백함수는 더이상 setTimeout의 함수가 아니게 된다. 그래서 catch 블록에서 캐치가 되지 않는다.

**콜백 헬**과 **에러처리가 곤란**하다는 문제를 극복하기 위해 ES6에 **프로미스**가 도입됐다.

## **프로미스**

**resolve :** 성공하면 실행하는 콜백함수

**reject :** 실패하면 실행하는 콜백함수

#### **프로미스의 상태정보**

**pending :** 비동기 처리가 아직 수행되지 않은 상태

**fulfilled :** 비동기 처리가 수행된 상태(성공)

**rejected :** 비동기 처리가 수행된 상태(실패)

#### **프로미스 후속처리 메서드**

**Promise.prototype.then :** 두개의 콜백함수를 인수로 전달받는다.(resolve가 반환됐을 때와 reject가 반환됐을 때)

**Promise.prototype.catch :** 프로미스가 reject상태인 경우에만 호출

**Promise.prototype.finally :**  한 개의 콜백 함수를 인수로 전달받는다. 성공여부와 상관없이 무조건 한번 호출

#### **프로미스의 에러처리**

then의 **두번째 콜백함수**에서 처리하거나 **catch문**을 사용해 처리할 수 있다.

단, then메서드의 두번쨰 콜백함수는 첫번째 콜백함수에서 발생한 에러를 캐치하지 못하고, 코드가 복잡해진다.

**promise.then(에러).then(여기서는 첫번째 then의 에러를 잡지 못함)**

**catch메서드**는 모든 then 메서드를 호출한 이후에 호출하면 비동기 처리에서 발생한 에러(rejected상태) 뿐만 아니라 then메서드 **내부에서 발생한 에러까지 모두 캐치할 수 있다.**

따라서 **catch메서드에서 에러처리를 하는 것을 권장**한다.

#### **프로미스 체이닝**

then, catch, finally 후속처리 메서드는 **언제나 Promise를 반환하므로 연속적으로 호출할 수 있다.**

이를 프로미스 체이닝이라고 한다.

프로미스는 프로미스 체이닝을 통해 처리결과를 전달 받아 후속처리르 하므로 비동기 처리를 위한 콜백 패턴에서 발생하던 콜백헬이 발생하지 않는다.

다만 프로미스도 콜백 패턴을 사용하므로 콜백함수를 사용하지 않는 것은 아니다.

콜백패턴은 가독성이 좋지 않다.

이 문제는 ES8에서 도입된 **async/await**를 통해 해결할 수 있다.

**async/await**를 사용하면 후속처리 메서드 없이 마치 동기 처럼 프로미스가 처리 결과를 반환하도록 구현할 수 있다.

### **프로미스의 정적 메서드**

Promise는 주로 생성자 함수로 사용되지만 함수도 객체이므로 메서드를 가질 수 있다.

```
const reslovedPromise = Promise.resolve([1,2,3]);
resolvedPromise.then(console.log); //[1,2,3]

const resolvedPromise = new Promise(resolve => resolve([1,2,3]);
resolvedPromise.then(console.log); //[1,2,3]

//두 예제는 같은 동작을 한다.
```

Promise.all : 여러 개의 비동기 처리를 모두 병렬처리 할 때 사용한다. (처리순서가 보장됨 -> 첫번째가 늦어도 차례대로 저장), 하나라도 rejected가 되면 그 즉시 종료

Promise.race : 가장 먼저 fulfilled 상태가 된 Promise의 처리결과를 resolve

Promise.allSettled : 전달받은 Promise가 모두 settled 상태(fulfilled 또는 rejected상태)가 되면 처리결과를 배열로 반환

-> fulfilled면 status 프로퍼티와 처리결과를 나타내는 value 프로퍼티

-> rejected면 status 프로퍼티와 에러를 나태는 reason 프로퍼티

## **마이크로 태스크 큐**

```
setTimeout(()=>console.log(1),0);

Promise.resolve().then(() => console.log(2))
				.then(()=>console.log(3));
//출력결과
/*
2
3
1
*/
```

위 실행결과는 **1 -> 2 -> 3** 이 아닌 **2 -> 3 -> 1** 이다.

그 이유는 **프로미스 후속 처리 메서드의 콜백 함수**는 태스크 큐가 아니라 **마이크로 태스크 큐에 저장되기 때문이다**.

**마이크로 태스크 큐**에는 프로미스 후속 처리 메서드 콜백 함수가 일시 저장됨

**태스크 큐**에는 비동기 함수의 콜백 함수나 이벤트 핸들러가 일시 저장됨

#### **마이크로 태스크 큐는 태스큐 큐보다 우선순위가 높다.**

즉 이벤트 루프는 콜 스택이 비면

먼저 마이크로 태스크 큐에서 대기하고 있는 함수를 가져와 실행한다.

이후 마이크로 태스크 큐가 비면 태스크 큐에서 대기하고 있는 함수를 가져와 실행한다.

## **fetch**

fetch함수는 XMLHttpRequest 객체와 마찬가지로 HTTP 요청 전송 기능을 제공하는 클라이언트 사이드 WEB API 이다.

fetch는 XMLHttpRequest 객체보다 사용법이 간단하고,

프로미스를 지원해 비동기 처리를 위한 콜백 패턴의 단점에서 자유롭다.

fecth함수는 HTTP 응답을 나타내는 **Response 객체를 래핑한 Promise 객체를 반환**

#### **fetch함수에서는 ok 상태를 필수로 확인해야한다?**

fetch 함수가 반환하는 프로미스는 기본적으로 404 Not Found나 500 Internal ServerError와 같은 HTTP 에러가 발생해도 에러를 reject 하지 않고, 불리언 타입의 ok 상태를 false로 설정한 Response 객체를 resolve한다.

오프라인 등의 네트워크 장애나 CORS 에러에 의해 요청이 완료되지 못한 경우에만 프로미스를 reject한다.

**따라서 fetch함수를 사용할 때는 fetch함수가 반환한 프로미스가 resolve한 불리언 타입의 ok상태를 확인해 명시적으로 에러를 처리할 필요가 있다.**
